spec/message.abnf

```diff
@@ -1,21 +1,15 @@
 message = [s] *(declaration [s]) body [s]
 
-declaration = let s variable [s] "=" [s] "{" [s] expression [s] "}"
+declaration = let s variable [s] "=" [s] expression
 body = pattern
      / (selectors 1*([s] variant))
 
-pattern = "{" *(text / placeholder) "}"
-selectors = match 1*([s] selector)
-selector = "{" [s] expression [s] "}"
+pattern = "{" *(text / expression) "}"
+selectors = match 1*([s] expression)
 variant = when 1*(s key) [s] pattern
 key = nmtoken / literal / "*"
 
-placeholder = "{" [s] expression [s] "}"
-            / "{" [s] markup-start *(s option) [s] "}"
-            / "{" [s] markup-end [s] "}"
-
-expression = ((literal / variable) [s annotation])
-           / annotation
+expression = "{" [s] (((literal / variable) [s annotation]) / annotation) [s] "}"
```

Comment:
For my action item I am working on ABNF, but we might as well talk about it here. I was going to suggest:

```abnf
expression = "{" [s] (((literal / variable) [s annotation])
             / annotation) [s] "}"
annotation = (function / reserved) *(s option)
function = ":" name
reserved = (%21-%26 / %2a / %2b / %2d / %2f
           / %3b-%40 / %5e / %7e) name ; reserved sigil characters, specifically these: !"#$%&*+-/^~
```

I'm reluctant to embrace `+` and `-` being "function introducers". I would rather just have a single starter for the "thing we've been calling a function" and having another name for other types of function. If these are different constructs, they should parse to another production name. That can be accommodated by pushing it down a level:

```abnf
annotation = named *(s option)
named = function / reserved
function = ":" name
reserved = [!@#$%^&*+-<>?] name
```
 This would allow something to become unreserved by adding a named production:

```abnf
annotation = named *(s option)
named = function / reserved
function = ":" name
fragment = "#" name
reserved = [!@$%^&*+-<>?] name ; deleted # from here
```

Reply:
I suspect that ultimately anything that may have an argument and/or a bag of options could be represented by some sort of "function". The proposed shape of `reserved` encodes this as the only possible direction for the extension of the spec, so ultimately anything that would be later un-reserved could effectively be described as a "function introducer".

If we don't want to limit ourselves this way, then really the `reserved` construct ought to be defined on a higher level, allowing for more freedom (using this PR's syntax as a baseline):
```abnf
expression = "{" [s] (((literal / variable) [s annotation]) / annotation / reserved) [s] "}"
reserved = reserved-start reserved-body
reserved-start = "!" / "@" / "#" / "$" / "%" / "^" / "&" / "*" / "<" / ">" / "?"
reserved-body = *(reserved-char / reserved-escape / ("{" reserved-body "}") / literal)
reserved-char = %x0-5B         ; omit \
              / %x5D-7A        ; omit { | }
              / %x7E-D7FF      ; omit surrogates
              / %xE000-10FFFF
reserved-escape = backslash ( backslash / "|" / "{" / "}" )
```

Now, independently of how we might proceed with `reserved`, I do think that "open" and "close" placeholders are sufficiently common that they definitely should have their own top-level introducers `+` and `-`. This leaves us with at least 11 more to use for other concepts, should that prove desirable later on.